### ISSUE TITLE
`DT` 0.20 fills `..._state` fields with non-reproducible values in `shinytest` snapshots

### ISSUE THREAD

**Issue by bersbersbers** [rstudio/DT#944]
`app.R`:
```r
shiny::shinyApp(
  shiny::fluidPage(DT::dataTableOutput("dt")),
  function(input, output, session) {
    output$dt <- DT::renderDataTable(cars)
  }
)
```

In terminal:
```shell
mkdir -p tests/shinytest
```

In R:
```r
app <- shinytest::ShinyDriver$new(); Sys.sleep(1); app$snapshot()
```

In terminal:
```shell
grep -1 time tests/shinytest/snapshot-current/001.json 
```

Note that the output is something like
```
    "t_state": {
      "time": 1637006522257,
      "start": 0,
```
changing every single time one runs `shinytest`.

I cannot find anything useful neither in the `DT` nor in the `DataTable` source code. Is there an obvious fix that I am missing, like disabling some `state` recording?

This is a problem with current HEAD, but not with 0.19 where `t_state` is simply `null`.

```
> xfun::session_info('DT')
R version 4.1.2 (2021-11-01)
Platform: x86_64-suse-linux-gnu (64-bit)
Running under: openSUSE Leap 15.2

Locale:
  LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
  LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

Package version:
  base64enc_0.1.3   crosstalk_1.2.0   digest_0.6.28     DT_0.20           fastmap_1.1.0     graphics_4.1.2    grDevices_4.1.2   htmltools_0.5.2   htmlwidgets_1.5.4
  jquerylib_0.1.4   jsonlite_1.7.2    later_1.3.0       lazyeval_0.2.2    magrittr_2.0.1    methods_4.1.2     promises_1.2.0.1  R6_2.5.1          Rcpp_1.0.7       
  rlang_0.4.12      stats_4.1.2       utils_4.1.2       yaml_2.2.1       
> 
```

**Comment by bersbersbers** [comment]
I tried this without success:

```r
shiny::shinyApp(
  shiny::fluidPage(DT::dataTableOutput("t")),
  function(input, output, session) {
    output$t <- DT::renderDataTable(cars, options = list(stateSave = F))
  }
)
```

And this looks related:

> Note: As of 1.11.0 the datatables state is stored regardless of the stateSave option.

https://datatables.net/reference/option/stateSave

**Comment by bersbersbers** [comment]
This does not work, either :(

```r
shiny::shinyApp(
  shiny::fluidPage(DT::dataTableOutput("t")),
  function(input, output, session) {
    output$t <- DT::renderDataTable(
      cars, options = list(initComplete = DT::JS(
        "function(settings, json) {settings.oInstance.api().state.clear()}"
      )),
    )
  }
)
```

**Comment by bersbersbers** [comment]
This somewhat ugly workaround works!

```r
shiny::shinyApp(
  shiny::fluidPage(DT::dataTableOutput("t")),
  function(input, output, session) {
    shiny::snapshotPreprocessInput("t_state", function(value) {})
    output$t <- DT::renderDataTable(cars)
  }
)
```

**Comment by yihui** [comment]
The only change we made in [DT 0.20](https://github.com/rstudio/DT/releases/tag/v0.20) was that we upgraded the DataTables library from 1.10.20 to 1.11.3. It seems something was changed in DataTables, but I'm not sure what exactly caused this issue...

**Comment by bersbersbers** [comment]
> The only change we made in [DT 0.20](https://github.com/rstudio/DT/releases/tag/v0.20) was that we upgraded the DataTables library from 1.10.20 to 1.11.3. It seems something was changed in DataTables, but I'm not sure what exactly caused this issue...

See https://github.com/rstudio/DT/issues/944#issuecomment-969298100, it's a documented change in DataTables >=1.11.0

**Comment by yihui** [comment]
Okay. I'm surprised that even `options = list(stateSave = F)` doesn't work, and DataTables always stores the state. Perhaps we should patch it inside DT somewhere to clear the state when `stateSave = F`, e.g., after we initialize the table: https://github.com/rstudio/DT/blob/dddedd1a974ce7bfcefe8cc49f5ed1ab36cf8d9a/inst/htmlwidgets/datatables.js#L276 Unfortunately I don't have time for this task. Would you be able to help? Thanks!

**Comment by bersbersbers** [comment]
@yihui unfortunately, since I have found a workaround above, I almost certainly have more urgent tasks until at least the end of the year (and only a minimum of experience with `DataTables` or `DT`, anyway).

**Comment by yihui** [comment]
Got you. No worries!

**Comment by wch** [comment]
One way to address this is to filter out or modify the `time` part of the state:
https://github.com/rstudio/DT/blob/d741c40ebf86031922a211b26cab00b40a524a2c/R/shiny.R#L181-L186

**Comment by bersbersbers** [comment]
For the record, the workaround I posted in https://github.com/rstudio/DT/issues/944#issuecomment-969332857 seems to have stopped working with v0.21 (at least in my app). However, the fix in #950 has made it into v0.21 at the same time, so there is no issue here.
